Kms

hagabor opened the pull request:
Added KMS Service with GenerateDataKey function only (with parameters only for KeyId and NumberOfBytes).

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/k6-jslib-aws?pullRequest=6) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/k6-jslib-aws?pullRequest=6) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/k6-jslib-aws?pullRequest=6) it.</sub>

oleiade commented on src/internal/kms.ts:
```diff
@@ -0,0 +1,143 @@
+        this.name = 'KMSServiceError'
+        this.operation = operation
+    }
+}
```
```suggestion
}

/**
 *  KMSKeyLength describes possible key lenght values for KMS API data key operations.
 */
enum KMSKeyLength {
    KeySize256 = 32,
    KeySize512 = 64,
}
```

oleiade commented on src/internal/kms.ts:
```diff
@@ -0,0 +1,143 @@
+    }
+}
+
+// TODO: create a Tags type
```
```suggestion
```

oleiade commented on src/internal/kms.ts:
```diff
@@ -0,0 +1,143 @@
+
+// TODO: create a Tags type
+
+export class DataKeyResp {
+    CiphertextBlob: string
+    KeyId: string
+    Plaintext: string
+
+    constructor(
+        CiphertextBlob: string,
+        KeyId: string,
+        Plaintext: string
+    ){
+        this.CiphertextBlob = CiphertextBlob
+        this.KeyId = KeyId
+        this.Plaintext = Plaintext
+    }
+    static fromJSON(json: JSONObject) {
+        return new DataKeyResp(
+            json.CiphertextBlob as string,
+            json.KeyId as string,
+            json.Plaintext as string
+        )
+    }
+}
```
```suggestion
/**
 * Class representing a data key
 */
export class DataKey {
    /**
     * The Amazon Resource Name (key ARN) of the KMS key that encrypted the data key.
     */
    id: string

    /**
     * The (base64-encoded) encrypted copy of the data key.
     */
    ciphertextBlob: string

    /**
     * The plaintext data key.
     * Use this data key to encrypt your data outside of KMS. Then, remove it from memory as soon as possible.
     */
    plaintext: string

    constructor(CiphertextBlob: string, KeyId: string, Plaintext: string) {
        this.ciphertextBlob = CiphertextBlob
        this.id = KeyId
        this.plaintext = Plaintext
    }

    static fromJSON(json: JSONObject) {
        return new DataKey(
            json.CiphertextBlob as string,
            json.KeyId as string,
            json.Plaintext as string
        )
    }
}
```

oleiade commented on src/internal/kms.ts:
```diff
@@ -0,0 +1,143 @@
+            'Content-Type': 'application/x-amz-json-1.1',
+        }
+    }
+    /**
+     * GenerateDataKey
+     * Used to generate data key with the KMS key defined
+     * @param {string} KeyId
+     * @param {number} NumberOfBytes
+     * @throws {KMSServiceError}
+     * @throws {InvalidSignatureError}
+     */
+    GenerateDataKey(KeyId: string, NumberOfBytes: number = 32): DataKeyResp | undefined {
+        const body = JSON.stringify({ KeyId: KeyId, NumberOfBytes : NumberOfBytes })
```
```suggestion

    /**
     * GenerateDataKey returns a unique symmetric data key for use outside of AWS KMS.
     *
     * This operation returns a plaintext copy of the data key and a copy that is encrypted under a symmetric encryption KMS key that you specify.
     * The bytes in the plaintext key are random; they are not related to the caller or the KMS key.
     * You can use the plaintext key to encrypt your data outside of AWS KMS and store the encrypted data key with the encrypted data.
     *
     * To generate a data key, specify the symmetric encryption KMS key that will be used to encrypt the data key.
     * You cannot use an asymmetric KMS key to encrypt data keys.
     *
     * Used to generate data key with the KMS key defined
     * @param {string} id - Specifies the symmetric encryption KMS key that encrypts the data key. Use its key ID, key ARN, alias name, or alias ARN.
     * @param {KMKeySize} size - Specifies the length of the data key in bytes. For example, use the value 64 to generate a 512-bit data key (64 bytes is 512 bits). Default is 32, and generates a 256-bit data key.
     * @throws {KMSServiceError}
     * @throws {InvalidSignatureError}
     * @returns {DataKey} - The generated data key.
     */
    GenerateDataKey(id: string, size: KMSKeySize = KMSKeySize.Size256): DataKey | undefined {
        const body = JSON.stringify({ KeyId: id, NumberOfBytes: size })
```

oleiade commented on src/internal/kms.ts:
```diff
@@ -0,0 +1,143 @@
+        })
+        this._handle_error('GenerateDataKey', res)
+
+        return DataKeyResp.fromJSON(res.json() as JSONObject)
```
```suggestion
        return DataKey.fromJSON(res.json() as JSONObject)
```

oleiade commented on src/internal/kms.ts:
```diff
@@ -0,0 +1,143 @@
+import { AWSError } from './error'
+import { AWSConfig } from './config'
+import { InvalidSignatureError, URIEncodingConfig } from './signature'
+import { v4 as uuidv4 } from 'uuid'
```
```suggestion
```

oleiade commented on src/internal/kms.ts:
```diff
@@ -0,0 +1,143 @@
+import { JSONArray, JSONObject } from 'k6'
```
```suggestion
import { JSONObject } from 'k6'
```

oleiade reviewed (changes requested):
Thank you so much for your contribution @hagabor 🎉 🙇🏻 Much appreciated!

This looks excellent indeed. I've left a bunch of suggestions that are mainly targeted towards improving type-safety, naming, documentation and hiding AWS internal naming conventions. Please let me know if you see potential further refinements, and incorporate them 🙇🏻

I will be testing this against our AWS setup and get back to you as soon as I have confirmed for myself that it works as expected.

Note that I'll be out of office for the upcoming week, and might be slower to repond in that time period 😉 

PS: Once we're good to go, I'll amend your PR with a regeneration of `.map` file as some type names will have changed 👍🏻   

Cheers!

oleiade commented:
PR Checklist:
- [ ] @oleiade test against actual KMS

oleiade commented:
I've tested the KMS feature you introduced @hagabor and everything seems to work as expected 🎉 Great job 🤝 

Please go ahead and integrate the suggested changes, and we should be good to merge this in preparation for a `v0.5.0` release. In order to stay productive, if I do not get an answer from you in the upcoming 7 days, I will go ahead and commit my suggestions and merge this so we don't stay blocked.

Cheers!

hagabor commented on src/internal/kms.ts:
```diff
@@ -0,0 +1,143 @@
+    operation: string
+
+    /**
+     * Constructs a SecretsManagerServiceError
```
```suggestion
     * Constructs a KMSServiceError
```

hagabor commented:
Dear Théo,

Sorry to responded with such delay. I'm going to integrate the suggested changes. 

brgds,
Gábor

hagabor commented on src/kms.ts:
```diff
@@ -0,0 +1,23 @@
+import { AWSConfig, InvalidAWSConfigError } from './internal/config'
+import {
+    KMSClient,
+    DataKeyResp,
```
```suggestion
    DataKey,
```

hagabor commented on src/kms.ts:
```diff
@@ -0,0 +1,23 @@
+    InvalidAWSConfigError,
+    // KMS
+    KMSClient,
+    DataKeyResp,
```
```suggestion
    DataKey,
```

hagabor commented:
Dear Théo,

I've added some changes. One place I've left comment from the initial SecretsManager service (what I've used as starting point). Also changed export, and import symbols. Now seems working. Please review if I can ask you. Thanks!  

Brgds,
Gábor

oleiade commented:
Hey @hagabor 

Thanks a lot for this, much appreciated 🙇🏻 
I'll merge this, and add end 2 end tests for it myself 👍🏻
